### PR TITLE
 expression: make in more powerful.

### PR DESCRIPTION
**Pull request by winoros**
Let `in` can handle more case, so we can make more expression able to calculate range.
PTAL @XuHuaiyu @zz-jason @hanfei1991

### CODE REVIEW

File: plan/expression_rewriter.go [hunk 1]
```diff
@@ -913,16 +913,23 @@ func (er *expressionRewriter) inToExpression(lLen int, not bool, tp *types.Field
...
+			if c, ok := args[i].(*expression.Constant); ok {
+				args[i] = expression.RefineConstantArg(er.ctx, c, opcode.EQ)
+			}
+		}
+	}
 	allSameType := true
-	for i := stkLen - lLen; i < stkLen; i++ {
-		if er.ctxStack[i].GetType().Tp != mysql.TypeNull && er.ctxStack[i].GetType().EvalType() != leftEt {
+	for _, arg := range args[1:] {
+		if arg.GetType().Tp != mysql.TypeNull && expression.GetAccurateCmpType(args[0], arg) != leftEt {
```
Comment (coocood) [inline plan/expression_rewriter.go]: Can we collect all the argument types and merge them to one type, then just cast all the expressions to this type and always use `in`?

Comment (coocood) [reply plan/expression_rewriter.go]: How about build DNF with different types?
`(intValLeft in (intA, intB, intC)) or (floatValLeft in (floatA, floatB))`?

Comment (winoros) [reply plan/expression_rewriter.go]: This is okay.

Comment (winoros) [reply plan/expression_rewriter.go]: But currently this is not very necessary to do this. This case can't push down to tikv, and can't be used to calculate range.

**Comment by coocood** [comment]
LGTM

File: util/ranger/range_test.go [hunk 2]
```diff
@@ -161,10 +161,12 @@ func (s *testRangerSuite) TestTableRange(c *C) {
 			resultStr:   "[(-inf,-1] [1,+inf)]",
 		},
 		// TODO: cast will change the extraction behavior of accessCondition.
```
Comment (hanfei1991) [inline util/ranger/range_test.go]: remove this comment ?

**Comment by winoros** [comment]
PTAL @hanfei1991 @zz-jason @XuHuaiyu

**Comment by winoros** [comment]
mysql-test needs to do some change. Don't merge for a while.

**Pull request merged**